job/src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java

```diff
@@ -272,6 +292,50 @@ private void copyLogs() {
             }
         }
     }
+
+    private String getLogPrefix(FlowNode node) {
+        BodyInvocationAction bodyInvocationAction = node.getAction(BodyInvocationAction.class);
+
+        if (bodyInvocationAction != null) {
+            BodyExecutionLabelAction bodyLabelAction = node.getAction(BodyExecutionLabelAction.class);
+
+            if (bodyLabelAction != null) {
+                return bodyLabelAction.getBodyLabel();
+            } else {
+                // This exec body does not have a body label. Stop now.
+                return null;
+            }
+        }
```

Comment:
Fair point. But it seems like I do need some way knowing when to stop searching up through the parents, otherwise I run into the situation where the `FlowEndNode` finds the `ParallelStep` as a parent (which seems very fishy to me, but I assume correct - [screenshot](https://www.dropbox.com/s/6npn2zqd3rk5w8y/Screenshot%202015-03-06%2015.20.10.png?dl=0)) and we end up thinking the end nodes are part of the parallel step, which is obviously not right.

e.g.

```
Started
Running: Allocate node : Start
Running on master in /var/folders/nw/tdwkgqsx1bj91gbwybt4d8540000gn/T/junit4918197128410133531/junit7632045101504281797/workspace/demo
Running: Allocate node : Body : Start
Running: Execute sub-workflows in parallel : Start
[a] Running: Parallel branch: a
[b] Running: Parallel branch: b
[a] Running: Print Message
[a] echo a
[a] Running: Print Message
[a] echo a
[a] Running: Execute sub-workflows in parallel : Body : End
[b] Running: Print Message
[b] echo b
[b] Running: Print Message
[b] echo b
[b] Running: Execute sub-workflows in parallel : Body : End
[a] Running: Execute sub-workflows in parallel : End
[a] Running: Allocate node : Body : End
[a] Running: Allocate node : End
[a] Running: End of Workflow
Finished: SUCCESS
```